File: panghae/PGS/알고리즘 고득점kit/깊이-너비 우선 탐색(DFS-BFS)/타겟 넘버.py
from collections import deque


def solution(numbers, target):
    answer = 0
    q = deque()

    q.append((0, 0))

    while q:
        num, index = q.popleft()

        if index == len(numbers):
            answer += 1 if num == target else 0
        else:
            q.append((num + numbers[index], index+1))
            q.append((num - numbers[index], index+1))

    return answer


print(solution([1, 1, 1, 1, 1], 3))
print(solution([4, 1, 2, 1], 4))

File: panghae/PGS/알고리즘 고득점kit/깊이-너비 우선 탐색(DFS-BFS)/타겟넘버.js
function solution(numbers, target) {
    let answer = 0;

    const q = [[0, 0]];
    let start = 0;
    let end = 1;

    while (start < end) {
        const [sum, index] = q[start++];

        if (index === numbers.length) {
            if (sum === target) {
                answer += 1;
            }
        } else {
            q[end++] = [sum + numbers[index], index + 1];
            q[end++] = [sum - numbers[index], index + 1];
        }
    }

    return answer;
}

console.log(solution([1, 1, 1, 1, 1], 3));
console.log(solution([4, 1, 2, 1], 4));

